dask/config.py

```diff
@@ -565,28 +565,52 @@ def get(
         try:
             result = result[k]
         except (TypeError, IndexError, KeyError):
-            if default is not no_default:
-                return default
-            else:
+            if default is no_default:
                 raise
+            return default
+
     return result
 
 
+def pop(key: str, default: Any = no_default, config: dict = config) -> Any:
+    """Like ``get``, but remove the element if found
+
+    See Also
+    --------
+    dask.config.get
+    dask.config.set
+    """
+    keys = key.split(".")
+    result = config
+    for i, k in enumerate(keys):
+        k = canonical_name(k, result)
+        try:
+            if i == len(keys) - 1:
+                return result.pop(k)
+            else:
+                result = result[k]
+        except (TypeError, IndexError, KeyError):
+            if default is no_default:
+                raise
+            return default
+
+
 def rename(aliases: Mapping, config: dict = config) -> None:
     """Rename old keys to new keys
 
     This helps migrate older configuration versions over time
     """
-    old = []
     new = {}
     for o, n in aliases.items():
-        value = get(o, None, config=config)
-        if value is not None:
-            old.append(o)
-            new[n] = value
-
-    for k in old:
-        del config[canonical_name(k, config)]  # TODO: support nested keys
+        try:
+            value = pop(o, config=config)
+        except (TypeError, IndexError, KeyError):
+            continue
+        warnings.warn(
+            f"dask config key {o!r} has been renamed to {n!r}",
+            FutureWarning,
+        )
```

Comment:
We have a separate `check_deprecations` method for handling deprecations here -- does this result in multiple warnings being emitted?

Reply:
Ah. I completely missed that.
This looks like a bit of a mess. dask/dask uses `check_deprecations`, while dask/distributed uses `rename`. The two systems are completely independent.

I'll rework this PR to unify them somehow.

Reply:
Upon further inspection:

- dask/dask uses `check_deprecations`, which works for `dask.config.set` but not for `~/.config/dask.yaml`
- dask/distributed uses `rename`, which works for  `~/.config/dask.yaml` but not for `dask.config.set`

I overhauled and unified the whole system.
Follow-up: https://github.com/dask/distributed/pull/8179